File: tree/1644_lowest_common_ancestor_of_a_binary_tree_ii/solution.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

/*
 * 1644. Lowest Common Ancestor of a Binary Tree II
 * https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree-ii/
 *
 * Given the root of a binary tree, return the LCA of two given nodes p and q.
 * If either p or q does NOT exist in the tree, return null.
 *
 * Difference from 236:
 *   - 236 guarantees both p and q exist in the tree.
 *   - 1644 does NOT — you must verify both were actually found before returning the LCA.
 *
 * Constraints: 1 <= nodes <= 10^4, all values unique, p != q
 */

function lowestCommonAncestor(root: TreeNode | null, p: TreeNode, q: TreeNode): TreeNode | null {
    let pExist = false;
    let qExist = false;

    function LCA(root: TreeNode | null, p: TreeNode, q: TreeNode): TreeNode | null {
        if (root == null) return null
        const left = LCA(root.left, p, q)
        const right = LCA(root.right, p, q)
        if (root === p) pExist = true
        if (root === q) qExist = true
        if (root === p || root === q) return root;
        if (left && right) {
            return root
        }
        return left ?? right
    }

    const res = LCA(root, p, q)
    return (pExist && qExist) ? res : null
};

File: tree/1644_lowest_common_ancestor_of_a_binary_tree_ii/solution.py
# 1644. Lowest Common Ancestor of a Binary Tree II
# https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree-ii/
#
# Same as 236, but p or q may NOT exist in the tree. Return None if either is missing.
# Key difference: cannot early-return on `root == p` — must traverse fully to confirm both exist.

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def lowestCommonAncestor(root, p, q):
    # TODO
    return None
